File: VLangCherry/server/internal/analizar/analizar.go
// Package analizar orquesta el pipeline completo: fuente .vch ->
// { errores, consola, simbolos, ast }. Entorno FRESCO por llamada.
package analizar

import (
	"fmt"

	"github.com/antlr4-go/antlr/v4"

	"vlangcherry/internal/ast"
	"vlangcherry/internal/parser"
	"vlangcherry/internal/runtime"
	"vlangcherry/internal/traductor"
)

type Resultado struct {
	Errores       []runtime.ErrorReporte `json:"errores"`
	Consola       string                 `json:"consola"`
	ConsolaLineas []string               `json:"consolaLineas"`
	Simbolos      []runtime.FilaSimbolo  `json:"simbolos"`
	AST           ast.Grafo              `json:"ast"`
	Dot           string                 `json:"dot"`
}

type oyenteErrores struct {
	*antlr.DefaultErrorListener
	errores *runtime.ListaErrores
	esLexer bool
}

func (o *oyenteErrores) SyntaxError(recognizer antlr.Recognizer, offendingSymbol interface{}, line, column int, msg string, e antlr.RecognitionException) {
	if o.esLexer {
		o.errores.Lexico(msg, line, column+1)
	} else {
		o.errores.Sintactico(msg, line, column+1)
	}
}

func Analizar(codigo string) Resultado {
	errores := runtime.NuevaListaErrores()

	entrada := antlr.NewInputStream(codigo)
	lexer := parser.NewVLangCherryLexer(entrada)
	lexer.RemoveErrorListeners()
	lexer.AddErrorListener(&oyenteErrores{errores: errores, esLexer: true})

	tokens := antlr.NewCommonTokenStream(lexer, antlr.TokenDefaultChannel)
	p := parser.NewVLangCherryParser(tokens)
	p.RemoveErrorListeners()
	p.AddErrorListener(&oyenteErrores{errores: errores, esLexer: false})

	arbolParse := p.Programa()

	var grafo ast.Grafo
	interprete := runtime.NuevoInterprete(errores)

	// Best-effort: aunque haya errores lexicos/sintacticos ya reportados,
	// se intenta traducir y ejecutar lo que se pudo parsear (igual criterio
	// que los demas proyectos del curso: recolectar TODOS los errores).
	func() {
		defer func() {
			if r := recover(); r != nil {
				errores.Semantico(fmt.Sprintf("error interno durante la ejecución: %v", r), 0, 0)
			}
		}()
		programa := traductor.Traducir(arbolParse)
		grafo = ast.ConstruirGrafo(programa)
		interprete.Interpretar(programa)
	}()

	consolaLineas := interprete.Consola()
	consolaTexto := ""
	for i, l := range consolaLineas {
		if i > 0 {
			consolaTexto += "\n"
		}
		consolaTexto += l
	}

	listaErrores := errores.Errores
	return Resultado{
		Errores:       listaErrores,
		Consola:       consolaTexto,
		ConsolaLineas: consolaLineas,
		Simbolos:      interprete.Simbolos(),
		AST:           grafo,
		Dot:           ast.ADot(grafo),
	}
}

File: CompInterpreter/server/src/analizar.js
/* ============================================================
   Orquestador: fuente .ci -> { errores, consola, simbolos, ast }.
   Conecta el parser Jison (lexico+sintactico), el interprete
   (semantico+ejecucion) y los reportes. Entorno FRESCO por llamada.
   ============================================================ */
const parserMod = require('./parser.js');
const { ListaErrores } = require('./interprete/errores');
const { Interprete } = require('./interprete/interprete');
const { construirGrafo, aDot } = require('./reportes/ast-grafo');

function obtenerParser() {
  // jison exporta { parser, Parser, parse, ... } o el parser directo
  if (parserMod.parser) return parserMod.parser;
  return parserMod;
}

function analizar(codigo) {
  const errores = new ListaErrores();
  const parser = obtenerParser();

  // yy compartido entre lexer y parser. El lexer empuja errores lexicos
  // directamente a este arreglo (misma referencia que errores.errores).
  parser.yy = { errores: errores.errores };

  // Recuperacion de errores sintacticos (modo panico via `error` en la gramatica).
  parser.yy.parseError = function (msg, hash) {
    let descripcion;
    if (hash && hash.token) {
      const encontrado = hash.text !== undefined && hash.text !== '' ? ('"' + hash.text + '"') : hash.token;
      descripcion = 'Se encontró ' + encontrado;
      if (hash.expected && hash.expected.length) {
        descripcion += ' y se esperaba: ' + hash.expected.slice(0, 6).join(', ');
      }
    } else {
      descripcion = msg;
    }
    const l = hash && hash.loc ? hash.loc.first_line : (hash ? hash.line + 1 : 0);
    const c = hash && hash.loc ? hash.loc.first_column + 1 : 0;
    errores.sintactico(descripcion, l, c);
    if (!hash || !hash.recoverable) {
      throw new Error(descripcion);   // lo atrapamos abajo
    }
  };
  // jison usa this.parseError; lo enlazamos tambien
  parser.parseError = parser.yy.parseError;

  let ast = null;
  try {
    ast = parser.parse(codigo);
  } catch (e) {
    // error sintactico no recuperable: los errores ya quedaron registrados.
    // Si no se registro ninguno (fallo inesperado), lo dejamos como sintactico.
    if (!errores.hay()) errores.sintactico(e.message || 'Error de análisis', 0, 0);
    ast = ast || null;
  }

  // Interprete (semantico + ejecucion). Best-effort aunque haya errores previos.
  const interprete = new Interprete(errores);
  if (ast) {
    try {
      interprete.interpretar(ast);
    } catch (e) {
      errores.semantico('Error interno durante la ejecución: ' + (e.message || e), 0, 0);
    }
  }

  const grafo = ast ? construirGrafo(ast) : { nodes: [], edges: [] };
  const listaErrores = errores.errores.slice().sort((a, b) =>
    (a.linea - b.linea) || (a.columna - b.columna));

  return {
    errores: listaErrores,
    consola: interprete.consola.join('\n'),
    consolaLineas: interprete.consola,
    simbolos: Array.from(interprete.simbolos.values()),
    ast: grafo,
    dot: aDot(grafo)
  };
}

module.exports = { analizar };
